Integration of polynomials over 2D polygons

ArifAhmed1995 opened the pull request:
Module for integrating uni/bivariate polynomials over 2-polytopes.

certik commented:
Hi @ArifAhmed1995. Thanks for starting to work on this. The Travis failure seems unrelated (#12697).

Would you mind creating some jupyter notebook with examples (and plots) how to use this?

When you create new PRs, or just want my feedback, please tag me using @certik, then I'll get an email and can comment on your work.

Let me know if you have any questions.

ArifAhmed1995 commented:
I'm close to getting the 2D prototype working. Once that's done I'll create the notebook and add it to the `examples` folder.

certik commented:
Perfect!

ArifAhmed1995 committed:
Complete prototype for 2D use case

ArifAhmed1995 committed:
Update 2D use case, add tests

certik commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,487 @@
+        s = 0
+        for i in point.values():
+            s += i ** 2
+        return s**(1/2)
```
This is a common mistake, see http://docs.sympy.org/dev/gotchas.html#python-numbers-vs-sympy-numbers, you have to use `S(1)/2`.

certik commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,487 @@
+                                    intersect[1] - x0[gens[1]]))
+            if is_vertex(intersect):
+                if isinstance(expr, Expr):
+                    value += distance_origin *\
```
space before `\`

certik reviewed (changes requested):
Also a test fails -- it could be perhaps caused by the `1/2` thing, per my other comment.

ArifAhmed1995 committed:
Change internal representations from dict to tuple/list

The dictionary representations for best origin and direction
parameter(w.r.t hyperplane) are changed to tuple/list form.

This is because of discrepancies arising due to {x: a, y: b}
and {y: b, x: a} referring to the same object. This change may
be temporary or permanent depending on how scalable and robust it
proves to be for the final 2D API and the 3D use case.

ArifAhmed1995 committed:
Rectify code, PEP8 fixes and add more tests

ArifAhmed1995 committed:
Revert change since list(set(x)) is arbitrary

ArifAhmed1995 commented:
@certik The tests should pass now. I've made the notebooks and have discussed the existing limitations of the methods therein.
We should discuss the final 2D API now so that I can implement it in this week (before 11th).

ArifAhmed1995 committed:
Remove implicit import statement

certik commented:
@ndotsu, here is @ArifAhmed1995 latest code.

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,448 @@
+    xl, yl = list(), list()
+
+    for vertex in poly.vertices:
+        xl.append(vertex.x)
```
It would be more efficient if you use `xl = list(map((lambda vertex: vertex.x), poly.vertices))`, similarly for `yl`.

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,448 @@
+    """
+    if dims is None:
+        if isinstance(expr, Expr):
+            dims = list(expr.free_symbols)
```
Can we have `tuple(expr.free_symbols)`? I don't think you are using mutability of `dims`.

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,448 @@
+        if isinstance(expr, Expr):
+            dims = list(expr.free_symbols)
+        else:
+            dims = [x, y]
```
`dims = tuple([x, y])`

gxyd commented:
I would say, you should use `var = S.Zero` instead of python int assignment `var = 0`.

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,448 @@
+    return ()
+
+
+def is_vertex(ent):
```
This sounds like it should be some class-level variable `is_Vertex` or `is_vertex`.

ArifAhmed1995 replied:
But how to do it for say `(2, 3)` or `(1, 2, 3)` ? Plus , this function may be expanded in the future.

gxyd replied:
Ah, leave it then. May be you can have it for any sequence (why just tuple?).

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,448 @@
+    expr: Denotes a polynomial(SymPy expression)
+    """
+    from sympy.plotting.plot import plot3d, plot
+    gens = list(expr.free_symbols)
```
no need for a `list`, can work with `gens = expr.free_symbols`.

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,448 @@
+        else:
+            poly_dict[1] = expr
+    else:
+        poly_dict = {0: expr}
```
better do `poly_dict[0] = expr`, instead of a complete variable re-assignment.

ArifAhmed1995 committed:
Improve code :

1. Use map function.
2. Use S.Zero instead of Python zero.
3. Use Tuple instead of List.

ArifAhmed1995 commented:
@gxyd Thanks for the review ! I've made all but one of the changes.

ArifAhmed1995 committed:
Add clockwise sort and minor change to reduction function

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -28,12 +30,15 @@ def polytope_integrate(poly, expr, dims=None):
         else:
             dims = (x, y)
 
+    if kwargs['clockwise'] is True:
```
You can use
```python
clockwise = kwargs.get('clockwise', True)
if clockwise:
    poly = ...
```

ArifAhmed1995 replied:
Ah yes. I've made the change locally. There's a deficiency in the clockwise sort so I'll push the total changes after correcting that.

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -335,17 +345,68 @@ def decompose(expr):
+        second = (b.x - center.x) * (b.x - center.x) + (b.y - center.y) * (b.y - center.y)
+        return -1 if first > second else 1
+
+    return Polygon(*sorted(vertices, key = cmp_to_key(compareTo)))
```
Looks better if you have `key=cmp_to_key(...)` https://www.python.org/dev/peps/pep-0008/#other-recommendations

certik commented on sympy/integrals/tests/test_intpoly.py:
```diff
@@ -0,0 +1,93 @@
+    assert polytope_integrate(Polygon(Point(0, 0), Point(0, 1),
+                                      Point(1, 1), Point(1, 0)), x * y) == 1/4
+    assert polytope_integrate(Polygon(Point(0, 3), Point(5, 3), Point(1, 1)),
+                              6*x**2 - 40*y) == float(-935)/3
```
Why is this a floating point number? We should try to compare to an exact number, i.e., `-S(935)/3`.

certik commented on sympy/integrals/tests/test_intpoly.py:
```diff
@@ -0,0 +1,93 @@
+    assert polytope_integrate(Polygon(Point(0, 0), Point(0, 2),
+                                      Point(4, 0)), 1, dims=(x, y)) == 4
+    assert polytope_integrate(Polygon(Point(0, 0), Point(0, 1),
+                                      Point(1, 1), Point(1, 0)), x * y) == 1/4
```
`1/4` -> `S(1)/4`

certik commented on sympy/integrals/tests/test_intpoly.py:
```diff
@@ -0,0 +1,93 @@
+    assert best_origin((2, 1), 3, l1, expr1) == (0, 3.0)
+    assert best_origin((2, 0), 3, l2, x ** 7) == (1.5, 0)
+    assert best_origin((0, 2), 3, l3, x ** 7) == (0, 1.5)
+    assert best_origin((1, 1), 2, l4, x ** 7 * y ** 3) == (0, 2.0)
```
Should return an exact number, not a floating point.

certik commented on sympy/integrals/tests/test_intpoly.py:
```diff
@@ -0,0 +1,93 @@
+
+    l1 = Segment2D(Point(0, 3), Point(1, 1))
+    l2 = Segment2D(Point(1.5, 0), Point(1.5, 3))
+    l3 = Segment2D(Point(0, 1.5), Point(3, 1.5))
```
1.5 -> S(3)/2

certik reviewed (commented):
In general, please do not use floating point numbers, but exact numbers, I commented on the places where I noticed it.

certik commented:
Also there seems to be a test failure at Travis:
```
File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/integrals/intpoly.py", line 30, in sympy.integrals.intpoly.polytope_integrate
Failed example:
    polytope_integrate(poly, expr)
Exception raised:
    Traceback (most recent call last):
      File "/opt/python/2.7.9/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest sympy.integrals.intpoly.polytope_integrate[6]>", line 1, in <module>
        polytope_integrate(poly, expr)
      File "/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages/sympy/integrals/intpoly.py", line 33, in polytope_integrate
        if kwargs['dims'] is None:
    KeyError: 'dims'
```

ArifAhmed1995 committed:
Following changes done in intpoly.py and test_intpoly:

intpoly():
1.decompose(): Returned dictionary also has constant with value 0.
2.integration_reduction(): Compute dot product directly --> only
                           one recursive call.
3.polytope_integrate(): Add support for hyperplane representation.

test_intpoly():
1. Add tests for hyperplane representation.
2. Add tests(two XFAILs) from the paper by Chen et al.(2015)

ArifAhmed1995 commented:
@certik I updated the code. The failing test runs fine locally :
```
============================= test process starts ==============================
executable:         /home/arif7/anaconda3/bin/python  (3.6.0-final-0) [CPython]
architecture:       64-bit
cache:              yes
ground types:       python 
random seed:        16576427
hash randomization: on (PYTHONHASHSEED=1314471289)

sympy/assumptions/tests/test_refine.py[8] ........                          [OK]

________________________________ slowest tests _________________________________
test_pow - Took 180.055 seconds
================= tests finished: 8 passed, in 184.17 seconds ==================
```

ArifAhmed1995 commented:
@certik As mentioned in my [blog post](https://arif7blog.wordpress.com/2017/06/17/week-3-reportjune-11-17-implementing-suggestions/) there are two major issues : 

* The current implementation of Polygons in the geometry module does not account for polygons with intersecting sides. I got the following error when running tests for polytopes with intersecting sides : 
```
sympy.geometry.exceptions.GeometryError: Polygon has intersecting sides.
```
Should I go ahead and modify the existing polygon class to accommodate this class of polygons ?

* The computation of `best origin` seems to be expensive. I'll try to optimize it by including some inexpensive preliminary checks and simply return first vertex if difference between exponents of variables is not too much. What that difference should be has to be decided by plotting some graphs and then set value to be that intersection point.

ArifAhmed1995 committed:
Complete Dynamic implementation

certik commented:
@ArifAhmed1995 I thought @ndotsu already commented on this --- but I don't see his answer now. Let me find it.

Regarding the remaining Travis failure (https://travis-ci.org/sympy/sympy/jobs/244146796), that is not related to this PR.

ArifAhmed1995 committed:
Add dynamic programming algorithm

ArifAhmed1995 commented:
@certik I've added the dynamic programming algorithm. This is a scalable technique and much faster than the standard one(200 - 300 times faster for higher power polynomials). I'll work on the intersecting polygon case now.
At the end of this week, I'll explain all the changes in a blog post. Maybe I can still fine-tune the new technique.

ArifAhmed1995 committed:
Remove standard algorithm

ArifAhmed1995 committed:
Major correction in DP algorithm

ArifAhmed1995 committed:
Implement use-case for computing many integrals given a maximum degree:

The user can enter a list of polynomials to integrate over a polytope
along with a maximum degree as input. Since all the monomial integrals
are calculated once and re-used, the process is faster as compared to
the case where each polynomial is computed individually.

certik commented:
@ArifAhmed1995 I found the email from @ndotsu, I am copying it here for reference:

Dear Arif,

A few answers and some further suggestions:

1. No need for intersecting polygons - complicates the issue and is also not a case that will be of widespread need/interest.  Can accommodate it down the road if there is time.

2. Yes, can skip `best_origin' for now ; in 3D it will be much more expensive and hence less attractive.  Can study it further down-the-road or come-up with some other approach that might tackle the high-exponent case. As you mention, for a square even if one has yx^100 the best origin will not work out for an edge that does not intersect the y-axis and hence it is not foolproof.  Also, rarely will one require to know the integral of just this monomial.  Moreover, for very large exponents and if the polygon is within the unit hypercube the value of the integral is miniscule - there are powerful methods to compute such integrals (see LattE code that we refer to in our paper).  For example, even in one dimension: \int_0^.5   x^100 = x^101/101 ~ 10^-33

3. Another option that would be worth exploring/implementing and useful is the integral of monomials up to degree p - also relevant to possibly #2 above.  Often in computational methods, we use polynomials of order (degree) p, i.e., it includes all monomials up to degree p. For example, in 2 dimensions if p = 2, then we have {1 , x, y, x^2, xy, y^2} as the monomials. In general,  poly = a_1 + a_2x + a_2 y + a_4 x^2 + a_5 xy + a_6 y^2.
In 2D, a degree p polynomials consists of a linear combination of (p+1)(p+2)/2 monomials whereas in 3D it is (p+1)(p+2)(p+3)/6.

So, the input expression can be a `vector' of monomials with the corresponding degrees-of-homogeneities, and the output is a vector consisting of the values of the integrals.  For example, if poly = [1  x  y x^2 xy y^2 ], then q = [0 1 1 2 2 2] or some such format.  If certain monomials are missing can indicate that it has coefficient a_i = 0.  For low to moderate-order polynomials (p \le m; need to know up to what m this can be efficient) the homogeneous integration algo. is greatly simplified:

* No recursion is needed. Start from {1}, then {x}, . . etc.  Store the intermediate values on each face/edge. 

* If so, then when you get to the term in 2D that has \int_F_i   x0 \cdot grad f ds then this is know since grad f is one degree less and it has been computed before.  For example, if f = x, then grad f = i + 0j and the integral become \int_F_i (x-component of x0) ds  but this is already known since \int_F_i ds has been found when you compute the integral of 1: \int_polygon 1 dA.  Same idea also extends to 3D - no recursion is needed and here it can prove to be a significant advantage.

In 3D, if p = 9, we would have 10.11.12/6 = 110x2 = 220 integrals to evaluate and hence unclear how fast it will be.  The pro is that no recursion is needed, it is a one-time cost and hence can easily evaluate a p-th degree polynomial again with different coefficients a_i without needing to recalculate.  The price one pays upfront will be worthwhile if repeated calculations are needed with different a_i. For low-order, i.e., say p <= 5, it is likely to be very efficient and attractive.

-suku.

certik commented:
It looks like everything important has been implemented, except possibly intersections.

Suku and I suggest to now wrap up what's in this PR and I will help you to get it into the state so that we can merge it into sympy. Once it is merged, the next step is to create a new PR that will implement the numerical quadrature on the polygon sides, which will be used to numerically check the exact answers provided by this PR.

And you should also start thinking about the 3D case, which we will start implementing soon.

The API for the 2D case looks good for now.

So the immediate work is to add docstrings for the functions and some doctests, and make sure all tests pass. Also the jupyter notebook should be moved to where all the other notebooks are in sympy. @ArifAhmed1995 once you do this, please let me know. We'll try to also review this PR by somebody else and then let's get it in.

If you have any questions, please let me know.

ArifAhmed1995 commented:
@certik All right. Will be finishing this part soon. I'll also write the blog post within a day.

ArifAhmed1995 committed:
Revert to standard algorithm and use dynamic for only multiple case.

On further testing the standard algorithm performed faster than the
dynamic one. However, if the terms of the polynomial are such that
every monomial's partial derivatives are present previously then dynamic
is faster. Therefore, it is used for the multiple case, that is when a
maximum degree is supplied and the values for a list of polynomials is to
be calculated.

smichr commented:
> everything important has been implemented, except possibly intersections.

If this is needed, it's not hard to fix this. For a numerical intersection, where `all(i.is_number for i in intersection_pt)`, one would just have to add all the intersections *in the order of increasing distance from A* when adding a Segment(A, B) to the polygon. So if C and D were such intersections when trying to add Segment(A, B), instead of adding Segment(A, D) the segments (A, C), (C,D) and (D, B) would need to be added.

ArifAhmed1995 commented:
@smichr I'm not aware of any way of constructing a  polygon given segments as input. Even so, breaking up `Segment(A, B)` into `Segment(A, C) + Segment(C, B)` wouldn't work because the line segments would be on same line and would collapse to `Segment(A, B)` again.

An initial solution would be to construct a graph with nodes as the vertices of polygon, use Fleury's Algorithm with an extra condition that moving from vertex A to B and then to C is not allowed if A, B and C are collinear. I'll try this out first. Do you think that this would be overkill and a much simpler way exists ?

certik commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,796 @@
+from __future__ import print_function, division
```
Let's add a module docstring describing what the module is for and what it does. Doesn't have to be long, just so that when somebody opens it, it is clear what it does.

ArifAhmed1995 replied:
Done.

certik commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,796 @@
+
+def polytope_integrate(poly, expr, **kwargs):
+    """Pre-processes the input data for integrating
+    univariate/bivariate polynomials over 2-Polytopes.
```
Isn't this the main function in this module that the user will call?

In that case, let's change it to something like:
```
"""
Integrates homogeneous functions over polytopes.

This function accepts the polytope in `poly` (currently only polygons are implemented) and the homogeneous function in `expr` (currently only univariate/bivariate polynomials are implemented) and returns the exact integral of `expr` over `poly`.

Parameters
...
```

Essentially imagine a user of sympy who has never heard of your module, or your function and let's say the user loads the documentation in a jupyter notebook using `polytope_integrate?`. Then just the docstring should provide enough information to understand what a general idea what this is doing, and then go into more and more details.

certik commented on sympy/integrals/tests/test_intpoly.py:
```diff
@@ -0,0 +1,155 @@
+    assert result_dict[expr2] == 13062161/27
+    assert result_dict[expr3] == 1946257153/924
+
+@XFAIL
```
Can you add a comment why this fails? Is it a bug, or not implemented yet?

ArifAhmed1995 replied:
This is issue being discussed above. I'll be working on this and try to get it implemented soon.

ArifAhmed1995 replied:
and yes, I'm doing on a different branch. Will submit separate PR for it.

certik replied:
I mean can you comment into the source code why it fails. Is it because these polynomials are interesecting? If so, then just put something like this into the code:
```python
# This test fails because the polynomials are interesecting and the code cannot handle it yet. This will be fixed in the future.
```
or something like that.

certik replied:
^^^ After this, I am fine to merge this PR.

ArifAhmed1995 replied:
Okay, done now.

certik reviewed (changes requested):
@ArifAhmed1995 great work. Thanks.

I went over this, and it looks very good. I left some minor comments.

Can you regenerate the notebook by doing "Kernel restart - erase output". That way it will remove all the (binary) output like the images, and only leaves a nice human readable json. This reduces size of the file. In general we should have a very good reason to commit binary information into the git repository, and I don't think this use case warrants it.

I suggest to only do the changes I pointed out, and do everything else as new PRs. Let's try to get this one in as a priority, and then work on other things.

ArifAhmed1995 committed:
Improve code :
1. Add module docstring.
2. Change main function docstring.
3. Add comment for XFAIL tests.
4. Modify decompose function, use parameter  as boolean
5. Modify hyperplane_parameters

Add notebook

smichr commented:
> I'm not aware of any way of constructing a  polygon given segments as input. Even so, breaking up Segment(A, B) into Segment(A, C) + Segment(C, B) wouldn't work because the line segments would be on same line and would collapse to Segment(A, B) again.

I meant segments loosely: pairs of coordinates denote segments of the polygon. Removal of collinear points must apply only to those not needed to define an intersection. Note that `pt` appears in the args of this polygon:

```
>>> pt = S.Half, S.Half
>>> Polygon((0,0),(1,0),pt,(0,1),(1,1),pt,(0,0))

Polygon(Point2D(0,0),Point2D(1,0),Point2D(1/2,1/2),Point2D(0,1),Point2D(1,1),Point2D(1/2,1/2))
```

certik commented:
@gxyd If you have time, would you mind reviewing this PR please? That would be a big help. It is ok to merge from me, except one very minor thing (https://github.com/sympy/sympy/pull/12673#discussion_r126161896). But since this is a new functionality, we wanted to know if you find something that we should fix before merging.

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,811 @@
+from sympy.core.function import diff
+
+from sympy.polys.polytools import LC
+from sympy.polys.polytools import gcd_list
```
The two imports could be merged as one, `from sympy.polys.polytools import LC, gcd_list`.

ArifAhmed1995 replied:
Done. Thanks for the review !

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,806 @@
+    return x0
+
+
+def decompose(expr, separate=0):
```
`separate` as `True/False` is more suitable and understandable, though the current one just works.

gxyd replied:
Using `separate=True` seems to return what `expr.args` would return, or does it differ?
```python

In [88]: e = x**2 + 3*y*x + x + y**5 + 2*x**5
In [90]: decompose(e, separate=True)
Out[90]: [x, y**5, x**2, 2*x**5, 3*x*y]

In [91]: e.args
Out[91]: (x, y**5, x**2, 2*x**5, 3*x*y)
```
Also though the name `separate` doesn't seem to fit, using `as_dict` instead sounds more like what it does.

certik replied:
Good point. @ArifAhmed1995 would you mind changing `separate=0` to `separate=False` and also to document in the docstring what `separate` does?

gxyd replied:
I think this `decompose` function could be improved, instead of what is done below could be done with already defined sympy functions like `degree_list`, `degree`.

For a given expression `s`, first extract all symbols. Using `sym = s.atoms(Symbol)` and then, using `degrees = [degree_list(i, *sym), i for i in s.args]` and then building the dictionary using `sum([i for i in degree])` for checking whether the element is already in dictionary and if yes then add the new element to the already existing term, otherwise add a new dictionary `(deg, monom)` pair.

I am not saying that the current logic is wrong, but rather this could be done using already existing functionality.

Apart from that I don't like seeing `if separate:` again and again (unless I don't understand the meaning of `separate`). You could at the end just have `list(poly_dict.values())` (when `separate=True`) would that do the same?

ArifAhmed1995 replied:
@certik Done.
@gxyd Didn't know about those methods in `polytools`. I've modified `decompose`. Thanks again for reviewing. Actually I feel `separate` is a better word than `as_dict` cause : `separate=True` would simply break-up(separate) the polynomial into constituent monomials and `separate=False` would group up similar degree polynomials.

certik commented on examples/notebooks/IntegrationOverPolytopes.ipynb:
```diff
@@ -0,0 +1,372 @@
+   "source": [
+    "### decompose(expr, separate=0) :\n",
+    "    expr : Uni/Bi-variate polynomial.\n",
+    "    separate(default : 0) : If separate is not 0 then return list of constituting monomials.\n",
```
Use `separate=False`, and change `0` to `False`

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,806 @@
+
+    def compareTo(a, b):
+        if a.x - center.x >= 0 and b.x - center.x < 0:
+            return -1
```
Unless there is some reason, I think it is better if sympified objects are returned, at other places as well in the code.

gxyd commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,806 @@
+        h = (point.x ** 2 + point.y ** 2) ** half
+    elif isinstance(point, dict):
+        s = 0
+        for i in point.values():
```
We can instead use `s = sum([i**2 for i in point.values()])`, it will also take care of empty `dict` (not need to worry about that).

gxyd commented:
@certik I will close down for now. Please go ahead and merge this don't hold off on me, I may not get time tomorrow.

ArifAhmed1995 commented on sympy/integrals/intpoly.py:
```diff
@@ -0,0 +1,781 @@
+    [((0, 1), 3), ((1, -2), -1), ((-2, -1), -3)]
+    """
+    params = []
+    vertices = list(poly.vertices)
```
Done. I should think more about optimizing code while writing it. `append` is a more expensive operation as compared to assignment.

certik commented:
Thanks @gxyd for the review and @ArifAhmed1995 for fixing it. If tests pass, this is now ok to merge with me.

I have one last request. The commit d76be087beba0e974a72db92d4478a2ce9582d27 adds a binary notebook (with binary images) into the git history. I would prefer if you can remove it from that commit.

In general, try to learn to create logical commits, so in the past, you should have created just one commit that adds the (binary) notebook, and not to lump it with some other changes in the code. Then it would be trivial to simply remove that commit now. The same thing when you removed the binary parts of the notebook --- you lumped that change with some other changes. 

You can edit the commits and remove it the binary notebook. Do you know how to do that? If not, you can either learn that (I think it's a very useful skill, you use `git rebase -i` and split the commits which touch the notebook, then you squash them), or you can also just squash all of the commits into just few, making sure the binary notebook is removed.

ArifAhmed1995 commented:
@certik Is it fine now ?

certik commented:
It's better, but the commit 6ddbad1ca551cd651414b0a0ef14fe40d8403069 still adds binary data. To make it simple, just remove the notebook from 6ddbad1ca551cd651414b0a0ef14fe40d8403069 and just add the notebook in 1bbf991bb4121060f1bce4d431226a89e70a21f2, without the binary images.

ArifAhmed1995 commented:
@certik This is done now, I hope. I will work on the intersecting polygon issue for one more day and then start on the 3D use-case. If the intersecting issue isn't solved today, I'll come back to it some other time.

certik reviewed (approved):
This looks good, +1 from me. Thanks for fixing everything @ArifAhmed1995.

ndotsu commented:
Arif,

The intersecting polygon case should work as is on the examples in the 
paper - no special treatment is needed so long as the code works for any 
nonconvex polygon (algorithm only `sees` the edges of the polygon).

-suku.


On 7/8/17 11:38 PM, Arif Ahmed wrote:
>
> @certik <https://github.com/certik> This is done now, I hope. I will 
> work on the intersecting polygon issue for one more day and then start 
> on the 3D use-case. If the intersecting issue isn't solved today, I'll 
> come back to it some other time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sympy/sympy/pull/12673#issuecomment-313882561>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AbwYKgmwNjhLpg35hTMu3__KLn89EGXRks5sL_bYgaJpZM4Nmzx0>.
>

-- 
N. Sukumar, Professor                      Work Phone: (530)754-6415
Department of Civil & Environmental Engg.  FAX       : (530)752-7872
3159 Ghausi Hall, One Shields Avenue       E-mail: nsukumar@ucdavis.edu
University of California                          *GO BLAZERS*
Davis, CA 95616             WEB:  http://dilbert.engr.ucdavis.edu/~suku

ArifAhmed1995 commented:
@ndotsu Yes, it will work once the `Polygon` class in SymPy can handle intersecting polygons without explicitly supplying those intersection points as input. I'm working on it right now.

ndotsu commented:
Okay, very good.  In case you have not already, do take a look at 
Section 5.4 in the paper for a discussion on this case.

On 7/9/17 11:02 AM, Arif Ahmed wrote:
>
> @ndotsu <https://github.com/ndotsu> Yes, it will work once the 
> |Polygon| class in SymPy can handle intersecting polygons without 
> explicitly supplying those intersection points as input. I'm working 
> on it right now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sympy/sympy/pull/12673#issuecomment-313907985>, or 
> mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AbwYKmPd92DdCXaNho_INj5wMWvlPGgFks5sMJdBgaJpZM4Nmzx0>.
>
-- 
N. Sukumar, Professor                      Work Phone: (530)754-6415
Department of Civil & Environmental Engg.  FAX       : (530)752-7872
3159 Ghausi Hall, One Shields Avenue       E-mail: nsukumar@ucdavis.edu
University of California                          *GO BLAZERS*
Davis, CA 95616             WEB:  http://dilbert.engr.ucdavis.edu/~suku

ArifAhmed1995 commented:
@smichr I've opened [#12931](https://github.com/sympy/sympy/pull/12931) to resolve the above issue of implicit intersections.
